src/main/java/frc/robot/subsystems/swerve/SwerveSubsystem.java

```diff
@@ -23,40 +23,79 @@
 import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Supplier;
 import org.littletonrobotics.junction.AutoLogOutput;
 import org.littletonrobotics.junction.Logger;
 
-public class Swerve extends SubsystemBase {
-  private static final double MAX_LINEAR_SPEED = Units.feetToMeters(14.5);
-  private static final double TRACK_WIDTH_X = Units.inchesToMeters(25.0);
-  private static final double TRACK_WIDTH_Y = Units.inchesToMeters(25.0);
-  private static final double DRIVE_BASE_RADIUS =
+public class SwerveSubsystem extends SubsystemBase {
+  // Drivebase constants
+  public static final double MAX_LINEAR_SPEED = Units.feetToMeters(14.5);
+  public static final double TRACK_WIDTH_X = Units.inchesToMeters(25.0);
+  public static final double TRACK_WIDTH_Y = Units.inchesToMeters(25.0);
+  public static final double DRIVE_BASE_RADIUS =
       Math.hypot(TRACK_WIDTH_X / 2.0, TRACK_WIDTH_Y / 2.0);
-  private static final double MAX_ANGULAR_SPEED = MAX_LINEAR_SPEED / DRIVE_BASE_RADIUS;
+  public static final double MAX_ANGULAR_SPEED = MAX_LINEAR_SPEED / DRIVE_BASE_RADIUS;
+  // Hardware constants
+  public static final int PigeonID = 0;
+
+  public static final ModuleConstants frontLeft =
+      new ModuleConstants("Front Left", 0, 1, 0, Rotation2d.fromRotations(0.0));
+  public static final ModuleConstants frontRight =
+      new ModuleConstants("Front Right", 2, 3, 1, Rotation2d.fromRotations(0.0));
+  public static final ModuleConstants backLeft =
+      new ModuleConstants("Back Left", 4, 5, 2, Rotation2d.fromRotations(0.0));
+  public static final ModuleConstants backRight =
+      new ModuleConstants("Back Right", 6, 7, 3, Rotation2d.fromRotations(0.0));
 
   public static final Lock odometryLock = new ReentrantLock();
   private final GyroIO gyroIO;
   private final GyroIOInputsAutoLogged gyroInputs = new GyroIOInputsAutoLogged();
-  private final Module[] modules = new Module[4]; // FL, FR, BL, BR
+  private final Module[] modules; // FL, FR, BL, BR
 
   private SwerveDriveKinematics kinematics = new SwerveDriveKinematics(getModuleTranslations());
   private Pose2d pose = new Pose2d();
   private Rotation2d lastGyroRotation = new Rotation2d();
 
-  public Swerve(
-      GyroIO gyroIO,
-      ModuleIO flModuleIO,
-      ModuleIO frModuleIO,
-      ModuleIO blModuleIO,
-      ModuleIO brModuleIO) {
+  public SwerveSubsystem(GyroIO gyroIO, ModuleIO... moduleIOs) {
     this.gyroIO = gyroIO;
-    modules[0] = new Module(flModuleIO, 0);
-    modules[1] = new Module(frModuleIO, 1);
-    modules[2] = new Module(blModuleIO, 2);
-    modules[3] = new Module(brModuleIO, 3);
+    modules = new Module[moduleIOs.length];
+    for (int i = 0; i < moduleIOs.length; i++) {
+      modules[i] = new Module(moduleIOs[i]);
+    }
+  }
+
+  /**
+   * Constructs an array of swerve module ios corresponding to the real robot.
+   *
+   * @return The array of swerve module ios.
+   */
+  public static ModuleIO[] getTalonFXModules() {
```

Comment:
I know this is a thing you've already got all over the code base, but I'd nudge you to reconsider naming things as getters that do allocations or work as a general piece of feedback. But specifically, this doesn't *get* something - it's effectively a call to new but isn't named like a factory method. `newTalonFXModule`?`createTalonFXModule`? `buildTalonFXModule`?